book/src/functional-testing.md

```diff
@@ -1,7 +1,84 @@
 # Functional Testing
 
-Some functional testing of your CSI driver can be done via the CLI mode of the [CSI `sanity`](https://github.com/kubernetes-csi/csi-test/tree/master/pkg/sanity) package.
-
-Drivers should also be functionally "end-to-end" tested while deployed in a Kubernetes cluster. Currently how to do this and what tests to run is left up to driver authors. In the future, a project (currently in development) aims to enable use of a pre-built `kubernetes/e2e/e2e.test` binary containing a standard set of Kubernetes CSI end-to-end tests to be imported and run by third party CSI drivers. This documentation will be updated with more information once that is ready to use.
+Drivers should be functionally "end-to-end" tested while deployed in a Kubernetes cluster. Previously, how to do this and what tests to run was left up to driver authors. Now, a standard set of Kubernetes CSI end-to-end tests can be imported and run by third party CSI drivers. This documentation specifies how to do so.
 
 The CSI community is also looking in to establishing an official "CSI Conformance Suite" to recognize "officially certified CSI drivers".  This documentation will be updated with more information once that process has been defined.
+
+# Kubernetes End to End Testing for CSI Storage Plugins
+
+Currently, [csi-sanity](https://github.com/kubernetes-csi/csi-test/tree/master/cmd/csi-sanity) exists to help test compliance with the CSI spec, but e2e testing of plugins is needed as well to provide plugin authors and users validation that their plugin is integrated well with specific versions of Kubernetes.
+
+## Setting up End to End tests for your CSI Plugin
+
+### Prerequisites:
+ * A Kubernetes v1.13+ Cluster
+ * [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl)
+
+There are two ways to run end-to-end tests for your CSI Plugin
+ 1) use [Kubernetes E2E Tests](https://github.com/kubernetes/kubernetes/tree/e84271ed8ad04c94ce3c0a6fecd9f8dff834e861/test/e2e/storage/external), by providing a DriverDefinition YAML file via a parameter. 
```

Comment:
For these higher level links where there's no specific line number referenced, I think it's fine to refer to the master for latest instructions

Reply:
done